src/coreclr/jit/flowgraph.cpp

```diff
@@ -52,6 +52,244 @@ static bool blockNeedsGCPoll(BasicBlock* block)
+//    PhaseStatus indicating what, if anything, was changed.
+//
+
+PhaseStatus Compiler::fgInsertClsInitChecks()
+{
+    if (!opts.OptimizationEnabled())
+    {
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    bool        modified = false;
+    BasicBlock* block;
+    BasicBlock* prevBb = nullptr;
+    for (block = fgFirstBB; block; block = block->bbNext)
+    {
+        if (!block->isRunRarely())
+        {
+            for (Statement* stmt : block->Statements())
+            {
+                for (GenTree* tree = stmt->GetTreeList(); tree != nullptr; tree = tree->gtNext)
+                {
+                    // we only need GT_CALL nodes with helper funcs
+                    if (!tree->IsCall())
+                    {
+                        continue;
+                    }
+
+                    GenTreeCall* call = tree->AsCall();
+                    if (call->gtCallType != CT_HELPER)
+                    {
+                        continue;
+                    }
+
+                    CorInfoHelpFunc helpFunc = eeGetHelperNum(call->gtCallMethHnd);
+                    if ((helpFunc != CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE) || (call->fgArgInfo->ArgCount() != 2))
+                    {
+                        continue;
+                    }
+
+                    GenTree* moduleIdArg = call->fgArgInfo->GetArgNode(0);
+                    GenTree* clsIdArg    = call->fgArgInfo->GetArgNode(1);
+
+                    if (!moduleIdArg->IsCnsIntOrI() || !clsIdArg->IsCnsIntOrI())
+                    {
+                        // Looks like moduleId or/and clsId were passed as indirect loads
+                        continue;
+                    }
+
+                    bool hasNotSupportedPreds = false;
+                    for (flowList* pred = block->bbPreds; pred != nullptr; pred = pred->flNext)
+                    {
+                        BasicBlock* predBlock = pred->getBlock();
+                        if ((predBlock->bbJumpKind != BBJ_NONE) && (predBlock->bbJumpKind != BBJ_ALWAYS) &&
+                            (predBlock->bbJumpKind != BBJ_COND))
+                        {
+                            hasNotSupportedPreds = true;
+                            break;
+                        }
+                        if (!BasicBlock::sameEHRegion(predBlock, block))
+                        {
+                            hasNotSupportedPreds = true;
+                            break;
+                        }
+                    }
+                    if (hasNotSupportedPreds)
+                    {
+                        continue;
+                    }
+
+                    if (prevBb == nullptr)
+                    {
+                        // We're going to emit a BB in front of fgFirstBB
+                        fgEnsureFirstBBisScratch();
+                        prevBb = fgFirstBB;
+                        if (prevBb == block)
+                        {
+                            continue;
```

Comment:
I am curious how often we bail out here...  any data?

Reply:
I found ~100 methods in PMI mode, but most of them belong to `System.Runtime.InteropServices.Variant` and `Microsoft.VisualBasic.CompilerServices.Symbols` classes. I guess it's not a big deal - just didn't want to mess with the scratchBB. Added a comment about it.